model/src/main/kotlin/PackageProvider.kt

```diff
@@ -0,0 +1,147 @@
+     * empty string.
+     */
+    val clearlyDefinedName: String = ""
+) {
+    COCOAPODS(
+        "^https?://cocoapods\\.org/pods/",
+        clearlyDefinedName = "cocoapods"
+    ),
+    CRATES_IO(
+        "^https?://crates\\.io/api/",
+        clearlyDefinedName = "cratesio"
+    ),
+    DEBIAN(
+        "^https?://(\\w+\\.)+debian.org/debian/",
+        clearlyDefinedName = "debian"
+    ),
+    GITHUB(
+        "^(https?|ssh)://(git@)?(\\w+\\.)?github\\.com/",
+        clearlyDefinedName = "github"
+    ),
+    GITLAB(
+        "^(https?|ssh)://(git@)?(\\w+\\.)?gitlab\\.com/",
+        clearlyDefinedName = "gitlab"
+    ),
+    GOLANG(
+        "^https?://go\\.googlesource\\.com/",
+        "^https?://proxy\\.golang\\.org/",
+        clearlyDefinedName = "golang"
+    ),
+    GRADLE_PLUGIN(
+        "^https?://plugins\\.gradle\\.org/m2/",
+        clearlyDefinedName = "gradleplugin"
+    ),
+    MAVEN_CENTRAL(
+        "^https?://repo\\.maven\\.apache\\.org/maven2/",
+        "^https?://repo1\\.maven\\.org/maven2/",
+        clearlyDefinedName = "mavencentral"
+    ),
+    MAVEN_GOOGLE(
+        "^https?://maven\\.google\\.com/",
+        clearlyDefinedName = "mavengoogle"
+    ),
+    NPM_JS(
+        "^https?://registry\\.npmjs\\.(com|org)/",
+        clearlyDefinedName = "npmjs"
+    ),
+    NUGET(
+        "^https?://api\\.nuget\\.org/.+\\.nupkg",
+        clearlyDefinedName = "nuget"
+    ),
+    PACKAGIST(
+        "^https?://repo\\.packagist\\.org/",
+        clearlyDefinedName = "packagist"
+    ),
+    PYPI(
+        "^https?://files\\.pythonhosted\\.org/packages/",
+        clearlyDefinedName = "pypi"
+    ),
+    RUBYGEMS(
+        "^https?://rubygems\\.org/gems/",
+        clearlyDefinedName = "rubygems"
+    ),
+
+    // The following have no matching ClearlyDefined provider.
+    ANDROID(
+        "^https?://android\\.googlesource\\.com/"
+    ),
+    APACHE_GIT(
+        "^https?://gitbox\\.apache\\.org/repos/.+\\.git",
+        "^https?://git-wip-us\\.apache\\.org/repos/.+\\.git"
+    ),
+    APACHE_SUBVERSION(
+        "^https?://svn\\.apache\\.org/repos/"
+    ),
+    BITBUCKET(
+        "^https?://bitbucket\\.org/"
+    ),
+    GOOGLE_CODE(
+        "^https?://\\w+\\.googlecode\\.com/"
+    ),
+    HACKAGE(
+        "^https?://hackage\\.haskell\\.org/package/"
+    ),
+    HASKELL(
+        "^https?://git\\.haskell\\.org/"
+    ),
+    JENKINS(
+        "^https?://repo\\.jenkins-ci\\.org/"
+    ),
+    JCENTER(
+        "^https?://jcenter\\.bintray\\.com/"
+    ),
+    JITPACK(
+        "^https?://jitpack\\.io/"
+    ),
+
+    // The following matches any (unknown) provider.
+    UNKNOWN(
+        ".*"
+    );
+
+    private val urlRegexes by lazy { urlPatterns.map { it.toRegex() } }
+
+    companion object {
+        fun get(url: String) = enumValues<PackageProvider>().find { provider ->
+            provider.urlRegexes.any { it.containsMatchIn(url) }
+        } ?: UNKNOWN
```

Comment:
UNKOWN is guaranteed to match, so `?: UNKOWN` should never happen. Would it make sense to express this with `!!` or rather `checkNotNull()` ?